File: 백준/Gold/10830. 행렬 제곱/행렬 제곱.py
import sys

n, b = map(int, sys.stdin.readline().rstrip().split())
p = 1000
A = [[int(x) % 1000 for x in sys.stdin.readline().rstrip().split()] for _ in range(n)]


def multiply(A, B):
    global n
    C = [[0 for _ in range(n)] for _ in range(n)]

    for i in range(n):
        for j in range(n):
            for k in range(n):
                C[i][j] += (A[i][k] * B[k][j]) % p
            C[i][j] %= p
    return C


def pow(A, exponent):
    if exponent == 1:
        return A

    temp = pow(A, exponent // 2)
    temp = multiply(temp, temp)

    if exponent % 2 == 1:
        temp = multiply(temp, A)

    return temp


result = pow(A, b)

for i in result:
    print(' '.join(map(str, i)))

File: 백준/Gold/10830. 행렬 제곱/행렬 제곱.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

	static int n;
	static long b;
	static int[][] A;
	static int p = 1000;

	public static void main(String[] args) throws IOException {

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine(), " ");

		n = Integer.parseInt(st.nextToken());
		b = Long.parseLong(st.nextToken());
		A = new int[n][n];

		for (int i = 0; i < n; i++) {
			st = new StringTokenizer(br.readLine(), " ");
			for (int j = 0; j < n; j++) {
				A[i][j] = Integer.parseInt(st.nextToken()) % p;
			}
		}

		int[][] result = pow(A, b);

		for (int i = 0; i < n; i++) {
			for (int j = 0; j < n; j++) {
				System.out.print(result[i][j] + " ");
			}

			System.out.println();
		}

	}

	public static int[][] pow(int[][] A, long exponent) {

		if (exponent == 1L) {
			return A;
		}

		int[][] temp = pow(A, exponent / 2);
		temp = multiply(temp, temp);

		if (exponent % 2 == 1) {
			temp = multiply(temp, A);
		}

		return temp;
	}

	public static int[][] multiply(int[][] A, int[][] B) {

		int[][] C = new int[n][n];
		for (int i = 0; i < n; i++) {
			for (int j = 0; j < n; j++) {
				for (int k = 0; k < n; k++) {
					C[i][j] += (A[i][k] * B[k][j]) % p;
				}
				C[i][j] %= p;
			}
		}

		return C;
	}
}
